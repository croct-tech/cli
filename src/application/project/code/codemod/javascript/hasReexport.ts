import * as t from '@babel/types';
import traverse from '@babel/traverse';
import {parse} from '@/application/project/code/codemod/parser';

export type ExportMatcher = {
    moduleName: string | RegExp,
    importName?: string | RegExp,
};

export function hasReexport(source: string | t.File, matcher: ExportMatcher): boolean {
    const ast = typeof source === 'string'
        ? parse(source, ['jsx', 'typescript'])
        : source;

    let found = false;

    traverse(ast, {
        // export * from 'something'
        ExportAllDeclaration: path => {
            const {node} = path;

            if (!t.isStringLiteral(node.source) || !matches(node.source, matcher.moduleName)) {
                return path.skip();
            }

            found = true;

            return path.skip();
        },

        // export {something} from 'something'
        // export {something as somethingElse} from 'something'
        ExportNamedDeclaration: path => {
            const {node} = path;

            if (!t.isStringLiteral(node.source) || !matches(node.source, matcher.moduleName)) {
                return path.skip();
            }

            if (matcher.importName === undefined) {
                found = true;

                return path.skip();
            }

            for (const specifier of node.specifiers) {
                if (t.isExportSpecifier(specifier) && matches(specifier.local, matcher.importName)) {
                    found = true;

                    return path.skip();
                }
            }

            return path.skip();
        },
    });

    return found;
}

function matches(value: t.Identifier|t.StringLiteral, matcher: string | RegExp): boolean {
    const name = t.isIdentifier(value) ? value.name : value.value;

    if (typeof matcher === 'string') {
        return name === matcher;
    }

    return matcher.test(name);
}
