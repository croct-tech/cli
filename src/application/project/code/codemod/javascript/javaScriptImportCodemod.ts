import {File} from '@babel/types';
import {Codemod, ResultCode} from '@/application/project/code/codemod/codemod';
import {ImportTransformer} from '@/application/project/code/codemod/globImportCodemod';
import {transformImports} from '@/application/project/code/codemod/javascript/utils/transformImports';

export type JavaScriptImportOptions = {
    transformer: ImportTransformer,
};

export class JavaScriptImportCodemod implements Codemod<File, JavaScriptImportOptions> {
    public async apply(file: File, options: JavaScriptImportOptions): Promise<ResultCode<File>> {
        const modified = await transformImports(file, options.transformer);

        return {
            result: file,
            modified: modified,
        };
    }
}
