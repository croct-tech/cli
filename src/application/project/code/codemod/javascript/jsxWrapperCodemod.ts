/* eslint-disable no-param-reassign -- False positives */
import * as t from '@babel/types';
import traverse from '@babel/traverse';
import {traverseFast} from '@babel/types';
import {ResultCode, Codemod, CodemodOptions} from '@/application/project/code/codemod/codemod';
import {addImport} from '@/application/project/code/codemod/javascript/utils/addImport';
import {getImportLocalName} from '@/application/project/code/codemod/javascript/utils/getImportLocalName';
import {AttributeType, createJsxAttributes} from '@/application/project/code/codemod/javascript/utils/createJsxProps';

type ComponentDeclaration = t.VariableDeclarator | t.FunctionDeclaration;
type DeclarationKind = t.ExportDefaultDeclaration['declaration'];
type ExpressionKind = t.Expression;
type JsxKind = t.JSXElement['children'][number];

type TargetChildren = {
    parent: t.JSXElement,
    index: number,
};

export type WrapperConfiguration<O extends CodemodOptions = CodemodOptions> = {
    wrapper: {
        component: string,
        module: string,
        props?: Record<string, AttributeType>,
    },
    targets?: {
        variable?: string,
        component?: string,
    },
    fallbackToNamedExports?: boolean,
    fallbackCodemod?: Codemod<t.File, O>,
};

export type WrapperOptions = CodemodOptions & {
    props?: Record<string, AttributeType>,
};

enum Transformation {
    APPLIED,
    NOT_APPLIED,
    ALREADY_APPLIED,
}

type WrapperInsertion = {
    result: Transformation,
    node: t.Expression,
};

/**
 * Wraps the exported component with a wrapper component.
 *
 * This transformation is useful for adding wrapper elements such as context providers.
 *
 * It attempts to wrap the default export first, and if not found, it can optionally
 * wrap named exports that return JSX elements depending on the configuration.
 */
export class JsxWrapperCodemod<O extends WrapperOptions = WrapperOptions> implements Codemod<t.File> {
    private readonly configuration: WrapperConfiguration<O>;

    public constructor(configuration: WrapperConfiguration<O>) {
        this.configuration = configuration;
    }

    public apply(input: t.File, options?: O): Promise<ResultCode<t.File>> {
        const importLocalName = getImportLocalName(input, {
            importName: this.configuration.wrapper.component,
            moduleName: this.configuration.wrapper.module,
        });

        const component = importLocalName ?? this.configuration.wrapper.component;
        const namedExports: t.ExportNamedDeclaration[] = [];

        let result = Transformation.NOT_APPLIED;

        traverse(input, {
            ExportDefaultDeclaration: path => {
                // Wrap the default export
                result = this.wrapDeclaration(path.node.declaration, component, input, options);

                return path.stop();
            },
            ExportNamedDeclaration: path => {
                // Collect named exports for fallback wrapping
                namedExports.push(path.node);

                return path.skip();
            },
        });

        if (result === Transformation.NOT_APPLIED && this.configuration?.fallbackToNamedExports === true) {
            // If the default export was not found, attempt to wrap named exports
            for (const namedExport of namedExports) {
                if (t.isFunctionDeclaration(namedExport.declaration)) {
                    // export function Component() { ... }
                    result = this.wrapBlockStatement(namedExport.declaration.body, component, input, options);

                    if (result === Transformation.APPLIED) {
                        break;
                    }
                }

                if (t.isVariableDeclaration(namedExport.declaration)) {
                    // export const Component = function() { ... };
                    const {declarations} = namedExport.declaration;

                    // Find the first variable declaration that is a variable,
                    // arrow function, or function expression
                    const modified = declarations.some(declaration => {
                        if (
                            t.isIdentifier(declaration.init)
                            || t.isArrowFunctionExpression(declaration.init)
                            || t.isFunctionExpression(declaration.init)
                        ) {
                            return this.wrapDeclaration(declaration.init, component, input, options)
                                === Transformation.APPLIED;
                        }

                        return false;
                    });

                    if (modified) {
                        result = Transformation.APPLIED;

                        break;
                    }
                }

                // export {Component as SomeComponent};
                for (const specifier of namedExport.specifiers ?? []) {
                    if (t.isExportSpecifier(specifier)) {
                        const declaration = this.findComponentDeclaration(input, specifier.local.name);

                        if (
                            declaration !== null
                            && t.isVariableDeclarator(declaration)
                            && t.isExpression(declaration.init)
                        ) {
                            result = this.wrapDeclaration(declaration.init, component, input, options);

                            if (result === Transformation.APPLIED) {
                                break;
                            }
                        }
                    }
                }

                if (result === Transformation.APPLIED) {
                    break;
                }
            }
        }

        if (result === Transformation.APPLIED && importLocalName === null) {
            const {body} = input.program;

            if (!t.isImportDeclaration(body[0])) {
                // Workaround to add a new line after the import statement
                body.unshift(t.emptyStatement());
            }

            addImport(input, {
                type: 'value',
                moduleName: this.configuration.wrapper.module,
                importName: this.configuration.wrapper.component,
            });
        }

        const fallbackCodemod = this.configuration?.fallbackCodemod;

        if (result === Transformation.NOT_APPLIED && fallbackCodemod !== undefined) {
            return fallbackCodemod.apply(input, options);
        }

        return Promise.resolve({
            modified: result === Transformation.APPLIED,
            result: input,
        });
    }

    /**
     * Wraps the declaration with the wrapper component.
     *
     * The declaration can be a function declaration, arrow function,
     * or variable declaration.
     *
     * @param node The declaration to wrap.
     * @param component The name of the component to use as the wrapper.
     * @param ast The AST representing the source code.
     * @param options The wrapper options.
     * @return the result of the transformation.
     */
    private wrapDeclaration(node: DeclarationKind, component: string, ast: t.File, options?: O): Transformation {
        if (t.isArrowFunctionExpression(node)) {
            if (t.isBlockStatement(node.body)) {
                return this.wrapBlockStatement(node.body, component, ast, options);
            }

            const insertion = this.insertWrapper(node.body, component, ast, options);

            if (insertion.result !== Transformation.APPLIED) {
                return insertion.result;
            }

            node.body = insertion.node;

            return Transformation.APPLIED;
        }

        if (t.isFunctionExpression(node) || t.isFunctionDeclaration(node)) {
            return this.wrapBlockStatement(node.body, component, ast, options);
        }

        if (t.isIdentifier(node)) {
            // Find the declaration of the component by name
            const declaration = this.findComponentDeclaration(ast, node.name);

            if (declaration !== null) {
                if (t.isVariableDeclarator(declaration)) {
                    const initializer = declaration.init ?? null;

                    if (initializer !== null) {
                        return this.wrapDeclaration(initializer, component, ast, options);
                    }
                } else {
                    return this.wrapBlockStatement(declaration.body, component, ast, options);
                }
            }
        }

        return Transformation.NOT_APPLIED;
    }

    /**
     * Wraps the block statement with the wrapper component.
     *
     * @param node The block statement to wrap
     * @param component The name of the component to use as the wrapper.
     * @param ast The AST representing the source code.
     * @param options The wrapper options.
     * @return the result of the transformation.
     */
    private wrapBlockStatement(node: t.BlockStatement, component: string, ast: t.File, options?: O): Transformation {
        const returnStatement = JsxWrapperCodemod.findReturnStatement(node);
        const argument = returnStatement?.argument ?? null;

        if (returnStatement !== null && argument !== null) {
            const insertion = this.insertWrapper(argument, component, ast, options);

            if (insertion.result !== Transformation.APPLIED) {
                return insertion.result;
            }

            returnStatement.argument = t.parenthesizedExpression(insertion.node);

            return Transformation.APPLIED;
        }

        return Transformation.NOT_APPLIED;
    }

    /**
     * Wraps the target element with the wrapper component.
     *
     * The given node may not be the target element itself, but a parent
     * that contains the target element. For example, the target element
     * may be a `{children}` expression inside a JSX element.
     *
     * @param node The node containing the target element.
     * @param component The name of the component to use as the wrapper.
     * @param ast The AST representing the source code.
     * @param options The wrapper options.
     * @return The result of the transformation.
     */
    private insertWrapper(node: ExpressionKind, component: string, ast: t.File, options?: O): WrapperInsertion {
        if (this.containsElement(node, component)) {
            return {
                result: Transformation.ALREADY_APPLIED,
                node: node,
            };
        }

        const target = this.findTargetChildren(ast, node);

        if (target !== null) {
            const {parent, index} = target;

            const children = [...parent.children ?? []];
            const child = children.splice(index, 1)[0] as JsxKind;

            target.parent.children = children.length === 0
                // If there is only one child, replace it with the wrapper
                ? [
                    t.jsxText('\n'),
                    this.wrapElement(child, component, options),
                    t.jsxText('\n'),
                ]
                // Otherwise, insert the wrapper at the target index
                : [
                    ...children.slice(0, index),
                    this.wrapElement(child, component, options),
                    ...children.slice(index),
                ];

            return {
                result: Transformation.APPLIED,
                node: node,
            };
        }

        if (
            t.isJSXText(node)
            || t.isJSXExpressionContainer(node)
            || t.isJSXSpreadChild(node)
            || t.isJSXElement(node)
            || t.isJSXFragment(node)
        ) {
            // Wrap the whole element if the target is not found
            return {
                result: Transformation.APPLIED,
                node: this.wrapElement(node, component, options),
            };
        }

        return {
            result: Transformation.NOT_APPLIED,
            node: node,
        };
    }

    /**
     * Wraps the given JSX element with the wrapper component.
     *
     * @param node The JSX element to wrap.
     * @param name The name of the component to use as the wrapper.
     * @param options The wrapper options.
     * @return The wrapped JSX element.
     */
    private wrapElement(node: JsxKind, name: string|undefined, options?: O): t.JSXElement {
        if (node.extra?.parenthesized === true) {
            node.extra.parenthesized = false;
        }

        return t.jsxElement(
            t.jsxOpeningElement(
                t.jsxIdentifier(name ?? this.configuration.wrapper.component),
                this.getProviderProps(options),
            ),
            t.jsxClosingElement(
                t.jsxIdentifier(name ?? this.configuration.wrapper.component),
            ),
            [
                t.jsxText('\n'),
                node,
                t.jsxText('\n'),
            ],
        );
    }

    /**
     * Returns the props for the wrapper component.
     *
     * @return The list of JSX attributes for the wrapper component.
     */
    private getProviderProps(options?: O): t.JSXAttribute[] {
        return createJsxAttributes({...this.configuration.wrapper.props, ...options?.props});
    }

    /**
     * Determines if the element contains the specified component.
     *
     * @param parent The parent element to search for the component.
     * @param name The name of the component to find.
     * @return true if the element contains the component, false otherwise.
     */
    private containsElement(parent: t.Node, name: string): boolean {
        let contains = false;

        traverseFast(parent, node => {
            if (
                !contains
                && t.isJSXOpeningElement(node)
                && t.isJSXIdentifier(node.name)
                && node.name.name === name
            ) {
                contains = true;
            }
        });

        return contains;
    }

    /**
     * Finds the declaration of the component by name.
     *
     * @param ast The AST representing the source code.
     * @param name The name of the variable or function to find.
     * @return The declaration of the component or null if not found.
     */
    private findComponentDeclaration(ast: t.File, name: string): ComponentDeclaration | null {
        let componentDeclaration: ComponentDeclaration | null = null;

        traverse(ast, {
            VariableDeclaration: path => {
                if (!t.isProgram(path.parent)) {
                    return path.skip();
                }

                for (const declaration of path.node.declarations) {
                    if (
                        t.isVariableDeclarator(declaration)
                        && t.isIdentifier(declaration.id)
                        && declaration.id.name === name
                    ) {
                        componentDeclaration = declaration;

                        return path.stop();
                    }
                }

                return path.skip();
            },
            FunctionDeclaration: path => {
                if (!t.isProgram(path.parent)) {
                    return path.skip();
                }

                const {id} = path.node;

                if (t.isIdentifier(id) && id.name === name) {
                    componentDeclaration = path.node;

                    return path.stop();
                }

                return path.skip();
            },
        });

        return componentDeclaration;
    }

    /**
     * Finds the target children of wrapper component.
     *
     * @param ast The AST representing the source code.
     * @param element The element to search for the target.
     * @return The parent element and index of the target children or null if not found.
     */
    private findTargetChildren(ast: t.File, element: t.Node): TargetChildren | null {
        let insertionPoint: TargetChildren | null = null;
        const {configuration} = this;

        if (configuration.targets?.variable === undefined && configuration.targets?.component === undefined) {
            // If no targets are specified, wrap the whole element
            return null;
        }

        traverse(ast, {
            enter: function enter(path) {
                const {node} = path;

                if (node !== element) {
                    return;
                }

                path.traverse({
                    JSXExpressionContainer: nestedPath => {
                        const {expression} = nestedPath.node;

                        if (
                            configuration.targets?.variable !== undefined
                            && t.isIdentifier(expression)
                            && expression.name === configuration.targets.variable
                        ) {
                            const parent = nestedPath.parent as t.JSXElement;

                            insertionPoint = {
                                parent: parent,
                                index: parent.children.indexOf(nestedPath.node),
                            };

                            return nestedPath.stop();
                        }
                    },
                    JSXElement: nestedPath => {
                        const {openingElement} = nestedPath.node;

                        if (
                            configuration.targets?.component !== undefined
                            && t.isJSXOpeningElement(openingElement)
                            && t.isJSXIdentifier(openingElement.name)
                            && openingElement.name.name === configuration.targets.component
                        ) {
                            if (nestedPath.parent !== null) {
                                const parent = nestedPath.parent as t.JSXElement;

                                insertionPoint = {
                                    parent: parent,
                                    index: parent.children.indexOf(nestedPath.node),
                                };
                            }
                        }
                    },
                });

                return path.stop();
            },
        });

        return insertionPoint;
    }

    /**
     * Finds the return statement in the block statement.
     *
     * @param body The block statement to search for the return statement.
     * @return The return statement or null if not found.
     */
    private static findReturnStatement(body: t.BlockStatement): t.ReturnStatement | null {
        let returnStatement: t.ReturnStatement | null = null;

        traverseFast(body, node => {
            if (returnStatement === null && t.isReturnStatement(node)) {
                returnStatement = node;
            }
        });

        return returnStatement;
    }
}
