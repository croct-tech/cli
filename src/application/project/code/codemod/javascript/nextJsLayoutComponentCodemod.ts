import * as t from '@babel/types';
import {Codemod, ResultCode} from '@/application/project/code/codemod/codemod';
import {addImport} from '@/application/project/code/codemod/javascript/utils/addImport';

export type LayoutComponentOptions = {
    typescript?: boolean,
};

export type LayoutComponentConfiguration = {
    provider: {
        component: string,
        module: string,
    },
};

export class NextJsLayoutComponentCodemod implements Codemod<t.File, LayoutComponentOptions> {
    private configuration: LayoutComponentConfiguration;

    public constructor(configuration: LayoutComponentConfiguration) {
        this.configuration = configuration;
    }

    public apply(input: t.File, options: LayoutComponentOptions = {}): Promise<ResultCode<t.File>> {
        const isTypescript = options.typescript ?? false;
        const {body} = input.program;

        const providerImport = addImport(input, {
            type: 'value',
            moduleName: this.configuration.provider.module,
            importName: this.configuration.provider.component,
        });

        const propsImport = isTypescript
            ? addImport(input, {
                type: 'type',
                moduleName: 'react',
                importName: 'PropsWithChildren',
            })
            : null;

        const reactNodeImport = isTypescript
            ? addImport(input, {
                type: 'type',
                moduleName: 'react',
                importName: 'ReactNode',
            })
            : null;

        const props = t.objectPattern([
            t.objectProperty(
                t.identifier('children'),
                t.identifier('children'),
            ),
        ]);

        if (propsImport !== null) {
            props.typeAnnotation = t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier(propsImport.localName)),
            );
        }

        const functionDeclaration = t.functionDeclaration(
            t.identifier('RootLayout'),
            [props],
            t.blockStatement([
                t.returnStatement(
                    t.parenthesizedExpression(
                        t.jsxElement(
                            t.jsxOpeningElement(
                                t.jsxIdentifier('html'),
                                [t.jsxAttribute(t.jsxIdentifier('lang'), t.stringLiteral('en'))],
                            ),
                            t.jsxClosingElement(t.jsxIdentifier('html')),
                            [
                                t.jsxText('\n'),
                                t.jsxElement(
                                    t.jsxOpeningElement(t.jsxIdentifier('body'), []),
                                    t.jsxClosingElement(t.jsxIdentifier('body')),
                                    [
                                        t.jsxText('\n'),
                                        t.jsxElement(
                                            t.jsxOpeningElement(t.jsxIdentifier(providerImport.localName), []),
                                            t.jsxClosingElement(t.jsxIdentifier(providerImport.localName)),
                                            [
                                                t.jsxText('\n'),
                                                t.jsxExpressionContainer(t.identifier('children')),
                                                t.jsxText('\n'),
                                            ],
                                        ),
                                        t.jsxText('\n'),
                                    ],
                                ),
                                t.jsxText('\n'),
                            ],
                        ),
                    ),
                ),
            ]),
        );

        if (reactNodeImport !== null) {
            functionDeclaration.returnType = t.tsTypeAnnotation(
                t.tsTypeReference(t.identifier(reactNodeImport.localName)),
            );
        }

        body.push(t.exportDefaultDeclaration(functionDeclaration));

        return Promise.resolve({
            modified: true,
            result: input,
        });
    }
}
