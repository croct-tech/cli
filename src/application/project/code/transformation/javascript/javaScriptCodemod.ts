import {File} from '@babel/types';
import {parse, print} from 'recast';
import {parse as babelParser} from 'recast/parsers/babel-ts.js';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/code/transformation/codemod';
import {Language} from '@/application/project/code/transformation/javascript/utils/parse';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<File, O>,
    languages: Language[],
};

export class JavaScriptCodemod<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<File, O>;

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const ast = parse(input, {
            parser: {
                parse: babelParser,
            },
        });

        const result = await this.codemod.apply(ast, options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result, {
                reuseWhitespace: false,
            }).code,
        };
    }
}
