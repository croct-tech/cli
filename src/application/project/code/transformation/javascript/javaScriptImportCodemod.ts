import {File} from '@babel/types';
import {Codemod, ResultCode} from '@/application/project/code/transformation/codemod';
import {ImportTransformer} from '@/application/project/code/transformation/globImportCodemod';
import {transformImports} from '@/application/project/code/transformation/javascript/utils/transformImports';

export type JavaScriptImportOptions = {
    transformer: ImportTransformer,
};

export class JavaScriptImportCodemod implements Codemod<File, JavaScriptImportOptions> {
    public async apply(file: File, options: JavaScriptImportOptions): Promise<ResultCode<File>> {
        const modified = await transformImports(file, options.transformer);

        return {
            result: file,
            modified: modified,
        };
    }
}
