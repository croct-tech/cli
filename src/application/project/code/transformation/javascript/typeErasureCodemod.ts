import {File, Comment, noop} from '@babel/types';
import type {VisitNodeObject, Node} from '@babel/traverse';
import {transformFromAstAsync, createConfigItem} from '@babel/core';
import bts from '@babel/plugin-transform-typescript';
import bsd from '@babel/plugin-syntax-decorators';
import {Codemod, ResultCode} from '@/application/project/code/transformation/codemod';
import {isTypescript} from '@/application/project/code/transformation/javascript/utils/isTypescript';

export type TypeErasureCodemodOptions = Record<string, never>;

export class TypeErasureCodemod implements Codemod<File, TypeErasureCodemodOptions> {
    public async apply(file: File): Promise<ResultCode<File>> {
        if (!isTypescript(file)) {
            return {
                modified: false,
                result: file,
            };
        }

        const babelTsTransform = createConfigItem([bts, {onlyRemoveTypeImports: true}]);
        const babelDecoratorSyntax = createConfigItem([bsd, {legacy: true}]);

        // When removing TS-specific constructs (e.g., interfaces),
        // ensure that any comments associated with those constructs are also removed.
        // Otherwise, leftover comments might reference nonexistent code.
        const removeComments: VisitNodeObject<unknown, Node> = {
            enter: function visit(nodePath) {
                const leadingComments = nodePath.node.leadingComments ?? nodePath.node.innerComments ?? null;

                if (leadingComments === null) {
                    return;
                }

                const preservedComments: Comment[] = [];

                for (const comment of leadingComments) {
                    const tokens = comment.loc?.tokens ?? [];
                    const tokenIndex = tokens.findIndex(
                        candidate => (
                            candidate.loc?.start === comment.loc?.start
                            && candidate.loc?.end === comment.loc?.end
                        ),
                    );

                    if (tokenIndex >= 0 && tokenIndex < tokens.length - 1) {
                        const commentToken = tokens[tokenIndex];
                        const nextToken = tokens.find(
                            (candidate, index) => (
                                index > tokenIndex
                                && !['CommentLine', 'CommentBlock'].includes(candidate.type)
                            ),
                        );

                        if (nextToken === undefined) {
                            continue;
                        }

                        const difference = nextToken.loc.start.line - commentToken.loc.end.line;

                        if (difference > 1) {
                            preservedComments.push(comment);
                        }
                    }
                }

                if (preservedComments.length > 0) {
                    const node = noop();

                    node.comments = preservedComments;

                    nodePath.insertBefore(node);
                }
            },
        };

        const result = await transformFromAstAsync(file, undefined, {
            plugins: [
                {
                    name: 'comment-remover',
                    visitor: {
                        Program: removeComments,
                        TSTypeAliasDeclaration: removeComments,
                        TSInterfaceDeclaration: removeComments,
                        TSDeclareFunction: removeComments,
                        TSDeclareMethod: removeComments,
                        TSImportType: removeComments,
                        TSModuleDeclaration: removeComments,
                    },
                },
                babelTsTransform,
                babelDecoratorSyntax,
            ],
            ast: true,
            configFile: false,
        });

        return {
            result: result?.ast ?? file,
            modified: (result?.ast ?? null) !== null,
        };
    }
}
