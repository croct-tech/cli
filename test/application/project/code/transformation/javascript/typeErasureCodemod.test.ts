import {resolve} from 'path';
import {loadFixtures} from '../fixtures';
import {JavaScriptCodemod} from '@/application/project/code/transformation/javascript/javaScriptCodemod';
import {TypeErasureCodemod} from '@/application/project/code/transformation/javascript/typeErasureCodemod';

describe('TypeErasureCodemod', () => {
    const scenarios = loadFixtures(resolve(__dirname, '../fixtures/ts-type-erasure'), {}, {});

    it.each(scenarios)('should correctly transform $name', async ({name, fixture}) => {
        const transformer = new JavaScriptCodemod({
            languages: ['typescript'],
            codemod: new TypeErasureCodemod(),
        });

        const output = await transformer.apply(fixture);

        expect(output.result).toMatchSnapshot(name);
    });
});
