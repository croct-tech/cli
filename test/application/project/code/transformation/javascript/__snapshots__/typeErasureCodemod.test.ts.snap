// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeErasureCodemod should correctly transform typescript.ts: typescript.ts 1`] = `
"export function addition(operands) {
    return operands.left + operands.right;
}"
`;

exports[`TypeErasureCodemod should correctly transform typescript-jsx.tsx: typescript-jsx.tsx 1`] = `
"export function Addition(operands) {
    return <div>{operands.left + operands.right}</div>;
}"
`;

exports[`TypeErasureCodemod should correctly transform typescript-line-comments.ts: typescript-line-comments.ts 1`] = `
"/**
 * This function takes two operands and returns their sum.
 *
 * @param operands
 */
export function addition(operands) {
    return operands.left + operands.right;
}"
`;

exports[`TypeErasureCodemod should correctly transform typescript-multiline-comments.ts: typescript-multiline-comments.ts 1`] = `
"/**
 * This function takes two operands and returns their sum.
 *
 * @param operands
 */
export function addition(operands) {
    return operands.left + operands.right;
}"
`;

exports[`TypeErasureCodemod should correctly transform typescript-preserve-comment.ts: typescript-preserve-comment.ts 1`] = `
"/**
 * Unrelated comment
 */


/**
 * This function takes two operands and returns their sum.
 *
 * @param operands
 */
export function addition(operands) {
    return operands.left + operands.right;
}"
`;

exports[`TypeErasureCodemod should correctly transform typescript-preserve-multiple-comments.ts: typescript-preserve-multiple-comments.ts 1`] = `
"// Unrelated comment
/**
 * Unrelated comment
 */


// Another unrelated comment
/**
 * Another unrelated comment
 */
/**
 * This function takes two operands and returns their sum.
 *
 * @param operands
 */
export function addition(operands) {
    return operands.left + operands.right;
}"
`;
